torch/autograd/function.py

```diff
@@ -186,10 +186,13 @@ def backward(ctx: Any, *grad_outputs: Any) -> Any:
         This function is to be overridden by all subclasses.
 
         It must accept a context :attr:`ctx` as the first argument, followed by
-        as many outputs did :func:`forward` return, and it should return as many
-        tensors, as there were inputs to :func:`forward`. Each argument is the
-        gradient w.r.t the given output, and each returned value should be the
-        gradient w.r.t. the corresponding input.
+        as many tensor outputs as the :func:`forward` returned (non tensor
```

Comment:
If we just ignore them, then what about having more constraint here that you must provide all Tensor outputs first, then as many non-Tensor objects that will be ignored?
That way the indexing of the Tensors properly match.

Beyond that, I am not 100% convinced about just ignoring these. It will make the forward/backward signature non-symetric which I am not sure I like..
cc @gchanan do you have an opinion on this?

Reply:
> If we just ignore them, then what about having more constraint here that you must provide all Tensor outputs first, then as many non-Tensor objects that will be ignored?
That way the indexing of the Tensors properly match.

I'm not sure if we need to worry about Tensor indexing. From an autograd point of view it only sees that the autograd function outputs tensors and no other values. 

> Beyond that, I am not 100% convinced about just ignoring these. It will make the forward/backward signature non-symetric which I am not sure I like..

I don't think the forward and backward signatures were symmetric earlier either right? You can write a forward function which takes as input 5 tensors and outputs only 3 tensors. In that case, your backward function would get as input 3 tensors, but it would return 5 tensors (for grads of the inputs).

Reply:
@albanD @gchanan Wondering if you had any more thoughts regarding this specific issue? Would like to get this change in soon to unblock some of our work on pipeline parallelism to support custom types :)

Reply:
I think we definitely want to keep this matching between the number of outputs of the forward and the number of inputs in the backward.

Reply:
And having all of them contiguous at the beginning doesn't feel necessary indeed, we can drop this idea.

Reply:
I updated the PR to make this symmetric, but had to touch `custom_function.cpp` for this since that controls how the autograd function metadata is setup. Although, I did it in a way such that it only impacts python custom functions and not cpp custom functions.